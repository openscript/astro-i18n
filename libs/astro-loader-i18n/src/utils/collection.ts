export function getAllUniqueKeys(obj: Record<string, unknown>, keys = new Set<string>()): Set<string> {
  Object.entries(obj).forEach(([key, value]) => {
    keys.add(key);
    if (value && typeof value === "object" && !Array.isArray(value)) {
      getAllUniqueKeys(value as Record<string, unknown>, keys);
    }
  });
  return keys;
}

function recursivePruneLocales(obj: Record<string, unknown>, locales: string[], locale: string) {
  const result: Record<string, unknown> = {};

  for (const [key, value] of Object.entries(obj)) {
    if (Array.isArray(value)) {
      result[key] = value.map((item) => {
        if (item && typeof item === "object" && !Array.isArray(item)) {
          return recursivePruneLocales(item as Record<string, unknown>, locales, locale);
        }
        return item;
      });
    } else if (value && typeof value === "object") {
      const valueAsRecord = value as Record<string, unknown>;
      const hasLocales = Object.keys(valueAsRecord).some((k) => locales.includes(k));
      let prunedValue: unknown | undefined = undefined;

      if (hasLocales) {
        prunedValue = valueAsRecord[locale] ?? undefined;
      } else {
        prunedValue = recursivePruneLocales(valueAsRecord, locales, locale);
      }

      result[key] = prunedValue;
    } else {
      result[key] = value;
    }
  }

  return result;
}

export function pruneLocales(obj: Record<string, unknown>, locales: string[], locale: string) {
  if (Object.keys(obj).find((key) => locales.includes(key))) throw new Error("Top-level locales are not allowed");
  if (Object.keys(obj).length === 0) return obj;

  return recursivePruneLocales(obj, locales, locale);
}
