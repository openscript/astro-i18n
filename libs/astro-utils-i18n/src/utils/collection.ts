export function getAllUniqueKeys(obj: Record<string, unknown>, keys = new Set<string>(), ignore?: boolean): Set<string> {
  Object.entries(obj).forEach(([key, value]) => {
    if (!ignore) keys.add(key);
    if (value && typeof value === "object") {
      if (Array.isArray(value)) {
        value.forEach((item) => {
          if (item && typeof item === "object") {
            // Recurse into objects and arrays within arrays
            if (Array.isArray(item)) {
              // Wrap array in an object to reuse the function and ignore its key
              getAllUniqueKeys({ array: item }, keys, true);
            } else {
              getAllUniqueKeys(item as Record<string, unknown>, keys);
            }
          }
        });
      } else {
        getAllUniqueKeys(value as Record<string, unknown>, keys);
      }
    }
  });
  return keys;
}

function recursivePruneLocales(obj: Record<string, unknown>, locales: string[], locale: string) {
  const result: Record<string, unknown> = {};

  for (const [key, value] of Object.entries(obj)) {
    if (Array.isArray(value)) {
      result[key] = value.map((item) => {
        if (item && typeof item === "object" && !Array.isArray(item)) {
          return recursivePruneLocales(item as Record<string, unknown>, locales, locale);
        }
        return item;
      });
    } else if (value && typeof value === "object") {
      const valueAsRecord = value as Record<string, unknown>;
      const hasLocales = Object.keys(valueAsRecord).some((k) => locales.includes(k));
      let prunedValue: unknown | undefined = undefined;

      if (hasLocales) {
        prunedValue = valueAsRecord[locale] ?? undefined;
      } else {
        prunedValue = recursivePruneLocales(valueAsRecord, locales, locale);
      }

      result[key] = prunedValue;
    } else {
      result[key] = value;
    }
  }

  return result;
}

export function pruneLocales(obj: Record<string, unknown>, locales: string[], locale: string) {
  if (Object.keys(obj).find((key) => locales.includes(key))) throw new Error("Top-level locales are not allowed");
  if (Object.keys(obj).length === 0) return obj;

  return recursivePruneLocales(obj, locales, locale);
}
