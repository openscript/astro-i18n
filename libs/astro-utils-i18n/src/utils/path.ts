const TRIM_SLASHES_PATTERN = /^\/|\/$/g;
const TRIM_RELATIVE_PATTERN = /^\.{0,2}\/|\/$/g;
const DOUBLE_POINTS_PATTERN = /\.{2}/g;
const DOUBLE_SLASH_PATTERN = /\/{2}/g;
const DIRNAME_PATTERN = /\/[^/]+$/g;
const INDEX_COMMON_TRANSLATION_ID = "index";
const createLocalePattern = (pathLocale: string) => new RegExp(`(?<=[./])${pathLocale}(?=[./])`, "i");

export function joinPath(...paths: Array<string | number | undefined>) {
  return paths.filter(Boolean).join("/");
}

/**
 * Resolves and joins multiple path segments into a single normalized path.
 *
 * This function trims leading and trailing slashes from string path segments
 * and ensures the resulting path starts with a single forward slash (`/`).
 * Non-string segments (e.g., numbers or `undefined`) are included as-is.
 *
 * @param paths - An array of path segments which can be strings, numbers, or `undefined`.
 * @returns A normalized path string starting with a forward slash.
 */
export function resolvePath(...paths: Array<string | number | undefined>) {
  const trimmedPaths = paths.map((path) => (typeof path === "string" ? path.replace(TRIM_SLASHES_PATTERN, "") : path));
  return `/${joinPath(...trimmedPaths)}`;
}

export const trimSlashes = (path: string) => {
  return path === "/" ? path : path.replace(TRIM_SLASHES_PATTERN, "");
};

export const trimRelativePath = (path: string) => {
  return path === "/" ? path : path.replace(TRIM_RELATIVE_PATTERN, "");
};

export const parseLocale = (path: string, locales: string[], defaultLocale: string) => {
  const localePattern = createLocalePattern(`(${locales.join("|")})`);
  const locale = path.match(localePattern)?.[0];

  return locale ? locale : defaultLocale;
};

export const createTranslationId = (path: string, locale?: string) => {
  if (locale) path = path.replace(createLocalePattern(locale), "");
  path = path.replace(DOUBLE_POINTS_PATTERN, ".");
  path = path.replace(DOUBLE_SLASH_PATTERN, "/");
  path = trimSlashes(path);

  return path === "" || path === "/" ? INDEX_COMMON_TRANSLATION_ID : path;
};

export const createContentPath = (path: string, base?: string | URL, locale?: string) => {
  let basePath = base ? (base instanceof URL ? base.pathname : base) : undefined;
  if (basePath) {
    basePath = trimRelativePath(basePath);
    const basePathIndex = path.indexOf(basePath);
    path = basePathIndex !== -1 ? path.slice(basePathIndex + basePath.length) : path;
  }
  path = createTranslationId(path, locale);
  if (path.includes(INDEX_COMMON_TRANSLATION_ID)) {
    path = path.replace(DIRNAME_PATTERN, "");
  }
  path = path.includes("/") ? path.replace(DIRNAME_PATTERN, "") : "";

  return path;
};
