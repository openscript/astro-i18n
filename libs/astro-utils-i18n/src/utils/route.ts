import { resolvePath } from "./path";

type Segments = Record<string, string | undefined>;
export type SegmentTranslations = Record<string, Segments>;
type RouteSegment = {
  spread: boolean;
  param: boolean;
  value: string;
};
type RoutePattern = RouteSegment[];

/**
 * Parses a route pattern into an array of route segments.
 *
 * @param routePattern is a string that templates the a route (e.g. /[...locale]/blog/[slug])
 */
export function parseRoutePattern(routePattern: string): RoutePattern {
  const segments = routePattern
    .split("/")
    .filter((segment) => segment !== "")
    .map((segment) => {
      if (segment.startsWith("[") && segment.endsWith("]")) {
        return {
          spread: segment.startsWith("[..."),
          param: true,
          value: segment.replace("[", "").replace("...", "").replace("]", ""),
        };
      }

      return {
        spread: false,
        param: false,
        value: segment,
      };
    });

  if (segments.length === 0) {
    return [{ spread: false, param: false, value: "/" }];
  }

  return segments;
}

/**
 * Constructs a full path by resolving a base path with a route pattern and corresponding segment values.
 *
 * @param routePattern - An array of route segments that define the structure of the route.
 *                        Each segment can either be a static value or a parameter.
 * @param segmentValues - An object mapping parameter names to their corresponding values.
 *                        These values are used to replace parameterized segments in the route pattern.
 * @param basePath - The base path to resolve the constructed route against.
 *
 * @returns The resolved path as a string.
 *
 * @throws {Error} If a required segment value for a parameterized route segment is missing
 *                 and the segment is not marked as a spread segment.
 */
export function buildPath(routePattern: RoutePattern, segmentValues: Segments, basePath: string) {
  return resolvePath(
    basePath,
    ...routePattern.map((segment) => {
      if (segment.param) {
        if (!segmentValues[segment.value] && !segment.spread) {
          throw new Error(`No segment value found for route segment "${segment.value}". Did you forget to provide it?`);
        }
        if (segmentValues[segment.value]?.includes("/") && !segment.spread) {
          throw new Error(
            `The segment value "${segmentValues[segment.value]}" for route segment "${segment.value}" contains a slash. Did you forget to add "..." to the route pattern?`
          );
        }

        return segmentValues[segment.value];
      }
      return `${segment.value}`;
    })
  );
}
