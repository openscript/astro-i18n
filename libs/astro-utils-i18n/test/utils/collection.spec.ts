import { describe, expect, it } from "vitest";
import { getAllUniqueKeys, pruneLocales } from "../../src/utils/collection";

describe("getAllUniqueKeys", () => {
  it("should return all unique keys of an object", () => {
    const result = getAllUniqueKeys({
      a: {
        b: {
          c: 1,
          d: 2,
        },
        e: 3,
      },
      b: 4,
      g: {
        a: 5,
        i: {
          c: 6,
          k: 7,
        },
      },
    });
    expect(result).toMatchSnapshot();
  });
});

describe("pruneLocales", () => {
  it("should prune locales with top level locale data as objects", () => {
    const result = () => {
      pruneLocales(
        {
          en: { a: 1, b: 2 },
          fr: { a: 3, b: 4 },
          de: { a: 5, b: 6 },
        },
        ["en", "fr", "de"],
        "en"
      );
    };
    expect(result).toThrowErrorMatchingSnapshot();
  });
  it("should prune locales with top level local data as strings", () => {
    const result = () => {
      pruneLocales(
        {
          en: "Hello",
          fr: "Bonjour",
          de: "Hallo",
        },
        ["en", "fr", "de"],
        "en"
      );
    };
    expect(result).toThrowErrorMatchingSnapshot();
  });
  it("should prune locales with no data", () => {
    const result = pruneLocales({}, ["en", "fr", "de"], "en");
    expect(result).toMatchSnapshot();
  });
  it("should prune locales with nested data", () => {
    const result = pruneLocales(
      {
        title: { en: "Title", fr: "Titre", de: "Titel" },
        nestedObject: {
          en: { a: 7, b: 8 },
          fr: { a: 9, b: 10 },
          de: { a: 11, b: 12 },
        },
        nestedArray: {
          en: [
            { a: 13, b: 14 },
            { a: 15, b: 16 },
          ],
          fr: [
            { a: 17, b: 18 },
            { a: 19, b: 20 },
          ],
          de: [
            { a: 21, b: 22 },
            { a: 23, b: 24 },
          ],
        },
        string: "normal string",
        number: 123,
        boolean: true,
        array: [1, 2, 3],
        object: { a: 1, b: 2 },
      },
      ["en", "fr", "de"],
      "en"
    );
    expect(result).toMatchSnapshot();
  });
});
