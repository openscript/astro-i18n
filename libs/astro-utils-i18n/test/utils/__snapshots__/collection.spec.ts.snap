// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getAllUniqueKeys > should return all unique keys of an object 1`] = `
Set {
  "a",
  "b",
  "c",
  "d",
  "e",
  "g",
  "i",
  "k",
}
`;

exports[`pruneLocales > should prune locales with nested data 1`] = `
{
  "array": [
    1,
    2,
    3,
  ],
  "boolean": true,
  "nestedArray": [
    {
      "a": 13,
      "b": 14,
    },
    {
      "a": 15,
      "b": 16,
    },
  ],
  "nestedObject": {
    "a": 7,
    "b": 8,
  },
  "number": 123,
  "object": {
    "a": 1,
    "b": 2,
  },
  "string": "normal string",
  "title": "Title",
}
`;

exports[`pruneLocales > should prune locales with no data 1`] = `{}`;

exports[`pruneLocales > should prune locales with top level local data as strings 1`] = `[Error: Top-level locales are not allowed]`;

exports[`pruneLocales > should prune locales with top level locale data as objects 1`] = `[Error: Top-level locales are not allowed]`;
